internals/overlord/planstate/manager.go

```diff
@@ -0,0 +1,192 @@
+package planstate
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/plan"
+)
+
+// LabelExists is the error returned by AppendLayer when a layer with that
+// label already exists.
+type LabelExists struct {
+	Label string
+}
+
+func (e *LabelExists) Error() string {
+	return fmt.Sprintf("layer %q already exists", e.Label)
+}
+
+type PlanManager struct {
+	state     *state.State
+	runner    *state.TaskRunner
+	pebbleDir string
+
+	planLock     sync.Mutex
+	plan         *plan.Plan
+	planHandlers []PlanChangeFunc
+}
+
+func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string) (*PlanManager, error) {
+	manager := &PlanManager{
+		state:     s,
+		runner:    runner,
+		pebbleDir: pebbleDir,
+	}
+
+	return manager, nil
+}
+
+// Load reads plan layers from the pebble directory, combine and validate the
+// final plan, and finally notifies registered managers up the plan update.
+func (m *PlanManager) Load() error {
+	m.planLock.Lock()
+	defer m.planLock.Unlock()
+	plan, err := plan.ReadDir(m.pebbleDir)
+	if err != nil {
+		return err
+	}
+	m.planChanged(plan)
+	return nil
+}
+
+// PlanChangeFunc is the type of function used by AddChangeListeners.
+type PlanChangeFunc func(p *plan.Plan)
+
+// AddChangeListeners adds f to the list of functions that are called whenever
+// the plan has changed. This method may not be called once the overlord state
+// engine has started.
+func (m *PlanManager) AddChangeListeners(f PlanChangeFunc) {
+	m.planHandlers = append(m.planHandlers, f)
+}
```

Comment:
*nit*: `AddChangeListener` (we only add one listener) -- or if you want to have plural, it should probably be  `AddToChangeListeners` (we add one thing to a list of listeners)

Also, "This method may not be called once the overlord state engine has started" -- is there a way to put this restriction into code (e.g. check if it has started, and fail somehow instead of silently ignoring it)?

Reply:
Fixed. Thank you. Resolving.